analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -197,89 +197,95 @@ private boolean canCombineWithLastRegisteredActivity(ActivityPayload payload, Ac
+		return lastRegisteredActivity.getEndTime().isBefore(getStartOfDay(payload.startTime, payload.userAnonymized));
 	}
 
-	private boolean isOnPrecedingDay(ActivityPayload payload, DayActivity lastRegisteredDayActivity)
+	private boolean hasLaterDay(ActivityDTO lastRegisteredActivity, ActivityPayload payload)
 	{
-		return payload.startTime.isBefore(lastRegisteredDayActivity.getStartTime());
+		return getStartOfDay(lastRegisteredActivity.getStartTime(), payload.userAnonymized).isAfter(payload.startTime);
 	}
 
-	private boolean isCrossDayActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized)
+	private boolean isCrossDayActivity(ActivityPayload payload)
 	{
-		return getStartOfDay(payload.startTime, userAnonymized).isBefore(getStartOfDay(payload.endTime, userAnonymized));
+		return getStartOfDay(payload.endTime, payload.userAnonymized).isAfter(payload.startTime);
 	}
 
-	private void addActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, GoalDTO matchingGoal,
-			DayActivityCacheResult dayActivityCacheResult)
+	private void addActivity(ActivityPayload payload, GoalDTO matchingGoal, ActivityCacheResult activityCacheResult)
 	{
-		Goal matchingGoalEntity = goalService.getGoalEntityForUserAnonymizedID(userAnonymized.getID(), matchingGoal.getID());
-		DayActivity updatedDayActivity = createNewActivity(dayActivityCacheResult.content, payload, userAnonymized,
-				matchingGoalEntity);
-		dayActivityRepository.save(updatedDayActivity);
-		if (dayActivityCacheResult.shouldUpdateCache())
+		Goal matchingGoalEntity = goalService.getGoalEntityForUserAnonymizedID(payload.userAnonymized.getID(),
+				matchingGoal.getID());
+		Activity addedActivity = createNewActivity(payload, matchingGoalEntity);
+		if (shouldUpdateCache(activityCacheResult, addedActivity))
 		{
-			cacheService.updateLastDayActivityForUser(updatedDayActivity);
+			cacheService.updateLastActivityForUser(payload.userAnonymized.getID(), matchingGoal.getID(),
+					ActivityDTO.createInstance(addedActivity));
 		}
 
 		if (matchingGoal.isNoGoGoal())
 		{
-			sendConflictMessageToAllDestinationsOfUser(payload, userAnonymized, updatedDayActivity.getLastActivity(),
-					matchingGoalEntity);
+			sendConflictMessageToAllDestinationsOfUser(payload, addedActivity, matchingGoalEntity);
 		}
 	}
 
-	private void updateActivityEndTime(ActivityPayload payload, UserAnonymizedDTO userAnonymized, GoalDTO matchingGoal,
-			DayActivityCacheResult dayActivity, Activity activity)
+	private void updateActivityEndTime(ActivityPayload payload, GoalDTO matchingGoal, ActivityCacheResult activityCacheResult)
 	{
-		assert userAnonymized.getID().equals(dayActivity.content.getUserAnonymized().getID());
-		assert matchingGoal.getID().equals(dayActivity.content.getGoal().getID());
-
+		DayActivity dayActivity = findExistingDayActivity(payload, matchingGoal.getID());
+		// TODO: are we sure that getLastActivity() gives the same activity? no concurrency issues?
```

Comment:
Good point.
To prevent concurrency, we should claim the user anonymized lock when analyzing activities.